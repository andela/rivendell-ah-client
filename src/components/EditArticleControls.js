import React, { Component } from 'react';
import { Button } from 'semantic-ui-react';

import CategoryInput from './CategoryInput';
import TagInput from './TagInput';


/**
 * This component manages the state of the input elements
 * to the edit Articles
 * @class EditControls
 * @returns {object} instance of an object
 * @param {object} e
 */
class EditControls extends Component {
  /**
   *This instantiates the state of this component
   * @param {object} props the properties passed to it as argument
   */
  constructor(props) {
    super(props);
    this.state = {
      currentValues: [],
      tags: [
        { key: 'English', text: 'English', value: 'English' },
        { key: 'French', text: 'French', value: 'French' },
        { key: 'Spanish', text: 'Spanish', value: 'Spanish' },
      ],
    };

    this.categories = [
      {
        name: 'OTHERS',
        subcategories: [
          {
            id: 1,
            name: 'OTHERS'
          }
        ]
      },
      {
        name: 'TECH',
        subcategories: [
          {
            id: 2,
            name: 'ENGINEERING'
          },
          {
            id: 3,
            name: 'SOFTWARE DEVELOPMENT'
          },
          {
            id: 4,
            name: 'ARTIFICIAL INTELLIGENCE'
          },
          {
            id: 5,
            name: 'BIOTECHNOLOGY'
          },
          {
            id: 6,
            name: 'NANO TECHNOLOGY'
          }
        ]
      },
    ];
  }


  /**
   *this function is called when a new field is inserted
    to the tags in while the user is creating articles
  @returns {void} this function performs an action and
  thus returns nothig
  @param {object} e the event handler object
   */
  handleAddition = (e, { value }) => {
    const { tags } = this.state;
    this.setState({
      tags: [{ key: value, text: value, value }, ...tags],
    });
  }

  /**
   *this function is called when the value of the
  tag changes
  @returns {void} this function performs an action and
  thus returns nothig
  @param {object} e the event handler object
   */
  handleChange = (e, { value }) => this.setState({ currentValues: value })


  /**
   * This returns a group of custom components that
   *@returns {JSX} a JSX to be rendered
   */
  render() {
    const { tags, currentValues } = this.state;
    return (
      <div>
        <TagInput
          options={tags}
          hadleAddition={this.handleAddition}
          handleChange={this.handleChange}
          currentValues={currentValues}
        />
        <CategoryInput categories={this.categories} />
        <Button primary content="Create Article" />
      </div>
    );
  }
}
export default EditControls;
