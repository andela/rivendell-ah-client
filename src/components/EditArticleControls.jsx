import React, { Component } from 'react';
import { Button, Input, Container, Loader }
  from 'semantic-ui-react';

import { connect } from 'react-redux';
import CategoryInput from './CategoryInput';
import TagInput from './TagInput';
import SmartEditor from './SmartEditor';
import createArticle from '../actions/createArticle';
import loadCategoriesAction from '../actions/loadCategories';
import ErrorMessagePortal from './ErrorMessagePortal';

/**
 * This component manages the state of the input elements
 * to the edit Articles
 * @class EditControls
 * @returns {object} instance of an object
 * @param {object} e
 */
export class EditArticleControls extends Component {
  /**
   *This instantiates the state of this component
   * @param {object} props the properties passed to it as argument
   */
  constructor(props) {
    super(props);

    const initialTagOptions = [
      { key: 'Art', text: 'Art', value: 'Art' },
      { key: 'Science', text: 'Science', value: 'Science' },
      { key: 'Life', text: 'Life', value: 'Life' },
    ];

    this.state = {
      tags: [],
      body: '',
      title: '',
      category: '',
      description: '',
      tagOptions: initialTagOptions,
    };

    this.switchToEditing = this.switchToEditing.bind(this);
    this.onBodyChange = this.onBodyChange.bind(this);
    this.submit = this.submit.bind(this);
    this.updateCategory = this.updateCategory.bind(this);
    this.updateTitle = this.updateTitle.bind(this);
    this.updateDescription = this.updateDescription.bind(this);
    this.updateTags = this.updateTags.bind(this);
  }


  /**
   * called immediately after the component has mounted
   *@returns {void} performs an action and returns nothing
   */
  componentDidMount() {
    const { loadCategories } = this.props;
    loadCategories();
  }

  /**
   *This method updates the article body based on what
   the user had inputed
   * @param {string} newArticleBody this contains html from
   * the smartEditor
   * @returns {void}
   */
  onBodyChange(newArticleBody) {
    this.setState({
      body: newArticleBody
    });
  }

  /**
   *this function is called when a new field is inserted
    to the tags in while the user is creating articles
  @returns {void} this function performs an action and
  thus returns nothig
  @param {object} e the event handler object
   */
  handleAddition = (e, { value }) => {
    const { tagOptions } = this.state;
    this.setState({
      tagOptions: [{ key: value, text: value, value }, ...tagOptions],
    });
  }

  /**
   *this function is called when the value of the
  tag changes
  @returns {void} this function performs an action and
  thus returns nothig
  @param {object} e the event handler object
   */
  updateTags(e, { value }) {
    this.setState({ tags: value });
  }

  /**
   *
   * Called when the createArticle button is clicked
   *@returns {void} return nothing
   */
  submit() {
    const { persistArticle } = this.props;
    const { title, body, tags, description, category } = this.state;

    console.log('Finally called submit');
    persistArticle({ title, description, body, tags, category });
  }

  /**
   *This updates the category once it is selected in the dropdown
   * makes changes to the title input
   * @param {object} e the event handler
   * @param {object} param1 an object taht contains the value
   * @returns {void} return nothing
   */
  updateCategory(e, { value }) {
    this.setState({
      category: value,
    });
  }

  /**
   *This updates the description once it is
   changed in the description input box
   * @param {object} e the event handler
   * @param {object} param1 an object taht contains the value
   * @returns {void} return nothing
   */
  updateDescription(e, { value }) {
    this.setState({
      description: value
    });
  }

  /**
   *This updates the title in the state when the user
   * makes changes to the title input
   * @param {object} e the event handler
   * @param {object} param1 an object taht contains the value
   * @returns {void} return nothing
   */
  updateTitle(e, { value }) {
    this.setState({
      title: value,
    });
  }


  /**
   * This dispatches an action to switch the state of this page
   * to editing mode
   * @returns {void} performs an action and returns void
   */
  switchToEditing() {
    const { clearErrors } = this.props;
    clearErrors();
  }

  /**
   * This returns a group of custom components that
   *@returns {JSX} a JSX to be rendered
   */
  render() {
    const { tagOptions } = this.state;
    const { categories, isLoading, errors } = this.props;

    const errorMessages = [];
    if (errors.title) {
      errorMessages.push(...errors.title);
    }

    if (errors.description) {
      errorMessages.push(...errors.description);
    }

    if (errors.body) {
      errorMessages.push(...errors.body);
    }
    return (
      <Container>

        <ErrorMessagePortal

          onHide={this.switchToEditing}
          open={!!errors.status}
          errorTitle="An error occured while making request"
          errorMessages={errorMessages}
        />
        <Input
          placeholder="Title"
          onChange={this.updateTitle}
          error={!!errors.title}
        />

        <Input
          placeholder="Description"
          onChange={this.updateDescription}
          error={!!errors.description}
        />
        <TagInput
          options={tagOptions}
          hadleAddition={this.handleAddition}
          handleChange={this.updateTags}
        />
        <CategoryInput
          categories={categories}
          onChange={this.updateCategory}
        />
        <Loader active={isLoading} inline content="Loading" />
        <SmartEditor
          onValueChange={this.onBodyChange}
        />

        <Button
          primary
          content="Create Article"
          onClick={this.submit}
          positive
        />
      </Container>
    );
  }
}


export const mapStateToProps = state => ({
  ...state.createArticleReducer,
  categories: state.loadCategoriesReducer.categories,

});
export const mapDispatchToProps = dispatch => ({
  persistArticle: (article) => {
    dispatch(createArticle(article));
  },
  loadCategories: () => {
    dispatch(loadCategoriesAction());
  },
  clearErrors: () => {
    dispatch({ type: 'CREATE_ARTICLE_EDITING' });
  }
});


export default connect(mapStateToProps,
  mapDispatchToProps)(EditArticleControls);
