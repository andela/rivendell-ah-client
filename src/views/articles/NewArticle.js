import React from 'react';
import BasicUserInfo from '../../components/BasicUserInfo';
import EditArticleControls from '../../components/EditArticleControls';


const NewArticle = () => (
  <div>
    <h2>Creating a new article</h2>
    <BasicUserInfo
      username="tempUser"
      image="https://i.ytimg.com/vi/3YJpUROvQAk/maxresdefault.jpg"
    />
    <EditArticleControls />
  </div>
);
export default NewArticle;
