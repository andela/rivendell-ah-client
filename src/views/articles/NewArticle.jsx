import React, { Component } from 'react';
import { Container } from 'semantic-ui-react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import EditArticleControls from '../../components/EditArticleControls';
import createArticleAction from '../../actions/createArticle';

/**
 * This component manages the state of the input elements
 * to the edit Articles
 * @class NewArticle
 * @returns {object} instance of an object
 */
export class NewArticle extends Component {
  /**
   * creates an instance of a NewArticle and binds the
   * methods to the this
   * @param {object} props contains the props sent from thd parent
   */
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  /**
   *
   * This function is called when the the user
   * tries to submit the article
   * @returns {void} performs an action and returns nothing
   * @param {object} article contains information about the
   * article
   */
  handleSubmit(article) {
    const { persistArticle } = this.props;
    persistArticle(article);
  }

  /**
   * this renders html
   * @returns {void} renders html an returns nothing
   */
  render() {
    const { storeState } = this.props;
    return (
      <Container>
        <EditArticleControls
          onSubmit={this.handleSubmit}
          {...storeState}
        />
      </Container>);
  }
}

NewArticle.propTypes = {
  persistArticle: propTypes.func,
  storeState: propTypes.shape({
    editing: propTypes.bool,
    isLoading: propTypes.bool,
    success: propTypes.bool,
    article: propTypes.shape,
    errors: propTypes.shape({
      description: propTypes.arrayOf(propTypes.string),
      title: propTypes.arrayOf(propTypes.string),
    }),
  })
};

NewArticle.defaultProps = {
  persistArticle: () => {},
  storeState: {}
};


export const mapStateToProps = state => ({
  storeState: state.createArticleReducer,
});

export const mapDispatchToProps = dispatch => ({
  persistArticle: article => dispatch(createArticleAction(article)),
});


export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NewArticle);
