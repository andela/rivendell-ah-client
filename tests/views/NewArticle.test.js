import { shallow } from 'enzyme';
import React from 'react';
import chai from 'chai';
import NewArticle from '../../src/views/articles/NewArticle';
import EditArticleControls from '../../src/components/EditArticleControls';

describe('testing NewArticle Copmonent', () => {
  it('should render an EditArticleControls component ', () => {
    const mountedObj = shallow(<NewArticle />);
    chai.expect(mountedObj.find(EditArticleControls).length)
      .to.equal(1);
  });

  it('should render an EditArticleControls component ', () => {
    const mountedObj = shallow(<NewArticle />);
    expect(mountedObj)
      .toMatchSnapshot();
  });
});
