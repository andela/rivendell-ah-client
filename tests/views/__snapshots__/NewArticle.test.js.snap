// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing NewArticle Copmonent should render an EditArticleControls component  1`] = `
<Container>
  <Connect(EditArticleControls) />
</Container>
`;
